src/Shared/SolutionConfiguration.cs

```diff
@@ -0,0 +1,127 @@
+{
+    internal sealed class SolutionConfiguration
+    {
+        private const string AttributeProject = "Project";
+
+        private const string AttributeAbsolutePath = "AbsolutePath";
+
+        // This field stores pre-cached project elements for project guids for quicker access by project guid
+        private readonly Dictionary<string, XmlElement> _cachedProjectElements = new Dictionary<string, XmlElement>(StringComparer.OrdinalIgnoreCase);
+
+        // This field stores pre-cached project elements for project guids for quicker access by project absolute path
+        private readonly Dictionary<string, XmlElement> _cachedProjectElementsByAbsolutePath = new Dictionary<string, XmlElement>(StringComparer.OrdinalIgnoreCase);
+
+        // This field stores the project absolute path for quicker access by project guid
+        private readonly Dictionary<string, string> _cachedProjectAbsolutePathsByGuid = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+        // This field stores the project guid for quicker access by project absolute path
+        private readonly Dictionary<string, string> _cachedProjectGuidsByAbsolutePath = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+        // This field stores the list of dependency project guids by depending project guid
+        private readonly Dictionary<string, List<string>> _cachedDependencyProjectGuidsByDependingProjectGuid = new Dictionary<string, List<string>>(StringComparer.OrdinalIgnoreCase);
+
+        public SolutionConfiguration(string xmlString)
+        {
+            XmlDocument? doc = null;
+
+            if (!string.IsNullOrEmpty(xmlString))
+            {
+                doc = new XmlDocument();
+                var settings = new XmlReaderSettings { DtdProcessing = DtdProcessing.Ignore };
+                using (XmlReader reader = XmlReader.Create(new StringReader(xmlString), settings))
+                {
+                    doc.Load(reader);
+                }
+            }
+
+            // Example:
+            //
+            // <SolutionConfiguration>
+            //  <ProjectConfiguration Project="{786E302A-96CE-43DC-B640-D6B6CC9BF6C0}" AbsolutePath="c:foo\Project1\A.csproj" BuildProjectInSolution="True">Debug|AnyCPU</ProjectConfiguration>
+            //  <ProjectConfiguration Project="{881C1674-4ECA-451D-85B6-D7C59B7F16FA}" AbsolutePath="c:foo\Project2\B.csproj" BuildProjectInSolution="True">Debug|AnyCPU<ProjectDependency Project="{4A727FF8-65F2-401E-95AD-7C8BBFBE3167}" /></ProjectConfiguration>
+            //  <ProjectConfiguration Project="{4A727FF8-65F2-401E-95AD-7C8BBFBE3167}" AbsolutePath="c:foo\Project3\C.csproj" BuildProjectInSolution="True">Debug|AnyCPU</ProjectConfiguration>
+            // </SolutionConfiguration>
+            //
+            if (doc?.DocumentElement != null)
+            {
+                foreach (XmlElement xmlElement in doc.DocumentElement.ChildNodes)
+                {
```

Comment:
This step is just to cache some stuff early? Is it always needed? I'm wondering if we could do it lazily. In particular, it looks like ProjectCacheService wasn't doing all of this before.

(Partially inspired by https://github.com/dotnet/msbuild/pull/8526, which got a big improvement by avoiding GetFileInfoNoThrow; I'm wondering if GetFullPathNoThrow might have a similar problem, though we don't actually have to crack open the file, so maybe not.)

Reply:
Yes, it's so that the file can be parsed just once. I'm not sure where laziness would add much since the `SolutionConfiguration` shouldn't be created at all if it won't be used.

Reply:
I guess where I got confused is that the previous ProjectCacheService version didn't do this early caching of parsed information step, yet it still looks like it just used fast access methods. Are things like `node.Attributes!["BuildProjectInSolution"]` secretly a slow linear search?

Reply:
It didn't do this caching, but as this is a 1-time thing for initialization, I'm not sure it really matters too much in practice.

I'll expose this part as a static method to avoid the extra allocation while still getting the code sharing.